pkg/status/combinedstatus-resolver.go

```diff
@@ -0,0 +1,470 @@
+
+	// NoteBindingResolution notes a binding resolution for status collection.
+	//
+	// 1. If `deleted` is true, the associated combinedstatus resolutions are
+	// removed from memory. The same is done if a resolution no longer requires
+	// status collection.
+	//
+	// 2. Excessive combinedstatus resolutions are removed if they are no
+	// longer associated with the binding.
+	//
+	// 3. For every workload object associated with one or more
+	// statuscollectors, a combinedstatus resolution is created/updated.
+	// The update may involve adding or removing statuscollectors, and changing
+	// the set of destinations associated with the binding.
+	//
+	// The function uses the statuscollector lister and the cached workstatus
+	// objects to update internal state. The given array is assumed to contain
+	// all objects that may be relevant to the binding.
+	//
+	// The returned set contains the identifiers of combinedstatus objects
+	// that should be queued for syncing.
+	NoteBindingResolution(bindingName string, bindingResolution binding.Resolution, deleted bool,
+		workStatuses []*workStatus,
+		statusCollectorLister controllisters.StatusCollectorLister) sets.Set[util.ObjectIdentifier]
+
+	// NoteStatusCollector notes a statuscollector's spec.
+	// The statuscollector is cached on the resolver's level, and is updated
+	// for every resolution it is involved with. The statuscollector is assumed
+	// to be valid and immutable.
+	//
+	// The given array is assumed to contain all objects that may be associated
+	// with any combinedstatus resolution.
+	// The array and its contents must not be mutated during this call.
+	//
+	// The returned set contains the identifiers of combinedstatus objects
+	// that should be queued for syncing.
+	NoteStatusCollector(statusCollector *v1alpha1.StatusCollector,
+		workStatuses []*workStatus) sets.Set[util.ObjectIdentifier]
+
+	// NoteWorkStatus notes a workstatus in the combinedstatus resolution
+	// associated with its source workload object.
+	//
+	// The returned set contains the identifiers of combinedstatus objects
+	// that should be queued for syncing.
+	NoteWorkStatus(workStatus *workStatus) sets.Set[util.ObjectIdentifier]
+
+	// ResolutionExists returns true if a combinedstatus resolution exists for
+	// the given binding name and workload object identifier.
+	ResolutionExists(bindingName string, objectIdentifier util.ObjectIdentifier) bool
+}
+
+// NewCombinedStatusResolver creates a new CombinedStatusResolver.
+func NewCombinedStatusResolver(celEvaluator *celEvaluator) (CombinedStatusResolver, error) {
+	celEvaluator, err := newCELEvaluator()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create CEL evaluator: %w", err)
+	}
+
+	return &combinedStatusResolver{
+		celEvaluator:             celEvaluator,
+		bindingNameToResolutions: make(map[string]map[util.ObjectIdentifier]*combinedStatusResolution),
+	}, nil
+}
+
+type combinedStatusResolver struct {
+	sync.RWMutex
+	celEvaluator *celEvaluator
```

Comment:
I think that it is conventional to list immutable and thread-safe fields before the mutex.

I am having trouble finding documentation asserting this, but I think that `cel-go/cel.Env` is thread-safe.

Reply:
Thanks